### ISSUE TITLE
obtaining Nan for energies while pretraining on AMD GPU cluster

### ISSUE THREAD

**Issue by bharath1996-hub** [google-deepmind/ferminet#99]
Hi all, I am trying to run small test case example of h4.py (which is in the test examples folder) on AMD PGU mi300machines. I am getting Nan for energies while pretraining itself. The behavior is same for all examples. when I only use CPU, no error shows up. Can someone please help me here? I am guessing there is a problem with jax but I don't know exactly what's the issue.

**packages** installed in the environment: 

absl-py==2.2.2
astroid==3.3.9
attrs==25.3.0
chex==0.1.89
cloudpickle==3.1.1
decorator==5.2.1
dill==0.4.0
distrax==0.1.5
dm-tree==0.1.9
docker==7.1.0
etils==1.12.2
-e git+https://github.com/google-deepmind/ferminet.git@c6f9db52fbb15cee01fd83c958230d27d61bfe70#egg=ferminet
flake8==7.2.0
folx @ git+https://github.com/microsoft/folx@a5f87028dd8648f967f5d27ed6b702e33614ec64
gast==0.6.0
h5py==3.13.0
immutabledict==4.2.1
importlab==0.8.1
iniconfig==2.1.0
isort==6.0.1
jax==0.5.0
jax-rocm60-pjrt==0.5.0
jax-rocm60-plugin==0.5.0
jaxlib==0.5.0
jaxtyping==0.3.2
Jinja2==3.1.6
kfac-jax @ git+https://github.com/deepmind/kfac-jax@0bfb0579a6681ea9c0cca3239fd7b05677c0455b
libcst==1.7.0
MarkupSafe==3.0.2
mccabe==0.7.0
ml_collections==1.1.0
ml_dtypes==0.5.1
msgspec==0.19.0
networkx==3.4.2
ninja==1.11.1.4
numpy==2.2.5
opt_einsum==3.4.0
optax==0.2.4
packaging==25.0
pandas==2.2.3
platformdirs==4.3.7
pluggy==1.5.0
pyblock==0.6
pycnite==2024.7.31
pycodestyle==2.13.0
pydot==3.0.4
pyflakes==3.3.2
pylint==3.3.6
pyparsing==3.2.3
pyscf==2.9.0
pytest==8.3.5
python-dateutil==2.9.0.post0
pytype==2024.10.11
pytz==2025.2
PyYAML==6.0.2
scipy==1.15.2
six==1.17.0
tabulate==0.9.0
tensorflow-probability==0.25.0
toml==0.10.2
tomlkit==0.13.2
toolz==1.0.0
typing_extensions==4.13.2
tzdata==2025.2
wadler_lindig==0.1.5
wrapt==1.17.2

Snippet of **output** is given below:

   pp:
    basis: ccecp-cc-pVDZ
    symbols: null
    type: ccecp
  pyscf_mol: null
  radius: 1.738
  set_molecule: _set_geometry
  states: 0
  type: 1
  units: angstrom
  use_pp: false

I0506 09:01:06.964008 22773395880832 __init__.py:92] No ROCm wheel installation found
INFO:2025-05-06 09:01:09,624:jax._src.xla_bridge:945: Unable to initialize backend 'tpu': INTERNAL: Failed to open libtpu.so: libtpu.so: cannot open shared object file: No such file or directory
I0506 09:01:09.624132 22773395880832 xla_bridge.py:945] Unable to initialize backend 'tpu': INTERNAL: Failed to open libtpu.so: libtpu.so: cannot open shared object file: No such file or directory
I0506 09:01:09.630630 22773395880832 train.py:405] Starting QMC with 4 XLA devices per host across 1 hosts.
converged SCF energy = -1.72985949273371  <S^2> = 1.0008981  2S+1 = 2.2368711
I0506 09:01:20.411187 22773395880832 train.py:605] No checkpoint found. Training new model.
E0506 09:01:26.414723  404910 buffer_comparator.cc:156] Difference at 32: 0, expected 511.347
E0506 09:01:26.414806  404910 buffer_comparator.cc:156] Difference at 33: 0, expected 494.674
E0506 09:01:26.414810  404910 buffer_comparator.cc:156] Difference at 34: 0, expected 505.968
E0506 09:01:26.414814  404910 buffer_comparator.cc:156] Difference at 35: 0, expected 512.798
.
.
.
2025-05-06 09:01:26.609200: E external/xla/xla/service/gpu/autotuning/gemm_fusion_autotuner.cc:1080] Results do not match the reference. This is likely a bug/unexpected loss of precision.
E0506 09:01:26.720371  404910 buffer_comparator.cc:156] Difference at 26624: 0, expected 61.998
E0506 09:01:26.720439  404910 buffer_comparator.cc:156] Difference at 26625: 0, expected 58.7483
E0506 09:01:26.720445  404910 buffer_comparator.cc:156] Difference at 26626: 0, expected 65.6573
E0506 09:01:26.720449  404910 buffer_comparator.cc:156] Difference at 26627: 0, expected 67.5546
E0506 09:01:26.720452  404910 buffer_comparator.cc:156] Difference at 26628: 0, expected 65.5742
E0506 09:01:26.720459  404910 buffer_comparator.cc:156] Difference at 26629: 0, expected 61.5136
E0506 09:01:26.720466  404910 buffer_comparator.cc:156] Difference at 26630: 0, expected 62.7582
E0506 09:01:26.720474  404910 buffer_comparator.cc:156] Difference at 26631: 0, expected 60.0108
E0506 09:01:26.720479  404910 buffer_comparator.cc:156] Difference at 26632: 0, expected 64.4018
E0506 09:01:26.720482  404910 buffer_comparator.cc:156] Difference at 26633: 0, expected 65.3498
2025-05-06 09:01:26.722590: E external/xla/xla/service/gpu/autotuning/gemm_fusion_autotuner.cc:1080] Results do not match the reference. This is likely a bug/unexpected loss of precision.
I0506 09:01:31.260658 22773395880832 pretrain.py:344] Pretrain iter 00000: nan 0.974365
I0506 09:01:31.267830 22773395880832 pretrain.py:344] Pretrain iter 00001: nan 0.974365
I0506 09:01:31.274075 22773395880832 pretrain.py:344] Pretrain iter 00002: nan 0.97168
I0506 09:01:31.280143 22773395880832 pretrain.py:344] Pretrain iter 00003: nan 0.973877
I0506 09:01:31.285818 22773395880832 pretrain.py:344] Pretrain iter 00004: nan 0.975586
I0506 09:01:31.291593 22773395880832 pretrain.py:344] Pretrain iter 00005: nan 0.968994
I0506 09:01:31.297721 22773395880832 pretrain.py:344] Pretrain iter 00006: nan 0.976807

**Comment by jsspencer** [comment]
Pretraining is supervised; the loss is a metric of the distance between the FermiNet orbitals and Hartree-Fock, rather than an energy. It is worrying that you get a NaN immediately in this stage. I have not seen this before. I do not have access to an AMD machine to try myself.  Do the JAX tests pass?

If you run on CPU only, does pretraining work? What about if you disable the XLA gemm autotuner?

**Comment by bharath1996-hub** [comment]
Yes the Jax tests pass in terms of jax accessing the GPUs. Only CPU runs and pretraining works. 

After some discussions with the AMD support staff, I got that the rocm-6.2.2 version is not compatible with jax and while latest versions support. We are working on the solution.

**Comment by bharath1996-hub** [comment]
Hi,

Update on this: I got the installation working. It is problem of installing jax and jax-lib as the rocm does not support the jax that gets installed while installing the repo.

One has to build the jax wheel from source and load the module and then clone the ferminet repository. one has to comment out the "`jax`" and "`jax-lib`" in setup.py file while installing other dependencies using `pip install -e .`
